packages/oidc-dpop-client-browser/src/dpop/tokenExchange.ts

```diff
@@ -0,0 +1,233 @@
+): value is { id_token: string } {
+  return value.id_token && typeof value.id_token === "string";
+}
+
+function hasRefreshToken(
+  value: { refresh_token: string } | Record<string, unknown>
+): value is { refresh_token: string } {
+  return value.refresh_token && typeof value.refresh_token === "string";
+}
+
+function hasTokenType(
+  value: { token_type: string } | Record<string, unknown>
+): value is { token_type: string } {
+  return value.token_type && typeof value.token_type === "string";
+}
+
+export type TokenEndpointResponse = {
+  accessToken: string;
+  idToken: string;
+  webid: string;
+  refreshToken?: string;
+  dpopJwk?: string;
+};
+
+export type TokenEndpointInput = {
+  grantType: string;
+  redirectUri: string;
+  code: string;
+  codeVerifier: string;
+};
+
+type WebidOidcIdToken = {
+  sub: string;
+  iss: string;
+  webid?: string;
+};
+
+function isWebidOidcIdToken(
+  token: WebidOidcIdToken | Record<string, unknown>
+): token is WebidOidcIdToken {
+  return (
+    (token.sub &&
+      typeof token.sub === "string" &&
+      token.iss &&
+      typeof token.iss === "string" &&
+      !token.webid) ||
+    typeof token.webid === "string"
+  );
+}
+
+/**
+ * Extracts a Webid from an ID token based on https://github.com/solid/webid-oidc-spec.
+ * The upcoming spec is still a work in progress.
+ *
+ * Note: this function does not implement the userinfo webid lookup yet.
+ * @param idToken
+ */
+async function deriveWebidFromIdToken(idToken: string): Promise<string> {
+  const decoded = await decodeJwt(idToken);
+  if (!isWebidOidcIdToken(decoded)) {
+    throw new Error(
+      `Invalid ID token: ${JSON.stringify(
+        decoded
+      )} is missing 'sub' or 'iss' claims`
+    );
+  }
+  if (decoded.webid) {
+    return decoded.webid;
+  }
+  if (!decoded.sub.match(/^https?:\/\/.+\..+$/)) {
+    throw new Error(
+      `Cannot extract WebID from ID token: the ID token returned by ${decoded.iss} has no webid claim, nor an IRI-like sub claim: [${decoded.sub}]`
+    );
+  }
+  return decoded.sub;
+}
+
+function validatePreconditions(
+  issuer: IIssuerConfig,
+  data: TokenEndpointInput
+): void {
+  if (
+    data.grantType &&
+    (!issuer.grantTypesSupported ||
+      !issuer.grantTypesSupported.includes(data.grantType))
+  ) {
+    throw new Error(
+      `The issuer [${issuer.issuer.toString()}] does not support the [${
+        data.grantType
+      }] grant`
+    );
+  }
+  if (!issuer.tokenEndpoint) {
+    throw new Error(
+      `This issuer [${issuer.issuer.toString()}] does not have a token endpoint`
+    );
+  }
+}
+
+function validateTokenEndpointResponse(
+  tokenResponse: Record<string, unknown>,
+  dpop: boolean
+): Record<string, unknown> & { access_token: string; id_token: string } {
+  if (!hasAccessToken(tokenResponse)) {
+    throw new Error(
+      `Invalid token endpoint response: ${JSON.stringify(
+        tokenResponse
+      )} is missing an access_token.`
```

Comment:
Suggest changing to:
`Invalid token endpoint response (missing the field 'access_token'): ${JSON.stringify(
        tokenResponse
      )}`